src/action/platform_info.rs

```diff
@@ -0,0 +1,271 @@
+            CannotGetOSType(ref error) => {
+                write!(fmt, "can't get OS type error: {}", error)
+            },
+            CannotGetLinuxRelease(ref error) => {
+                write!(fmt, "can't get Linux release info error: {}", error)
+            }
+        }
+    }
+}
+
+impl From<Error> for session::Error {
+
+    fn from(error: Error) -> session::Error {
+        session::Error::action(error)
+    }
+}
+
+/// A Response type for `GetPlatformInfo` action
+#[derive(Default)]
+pub struct Response {
+    /// Client platform information
+    system: Option<String>,
+    release_name: Option<String>,
+    version_id: Option<String>,
+    machine: Option<String>,
+    kernel_release: Option<String>,
+    fqdn: Option<String>,
+    architecture: Option<String>,
+    node: Option<String>
+}
+
+/// Function that converts raw C-strings into `String` type
+fn convert_raw_string(c_string: &[c_char]) -> String {
+    unsafe { 
+        String::from(CStr::from_ptr(c_string.as_ptr()).to_string_lossy().into_owned())
+    } 
+}
+
+/// Function that returns `Response` for Linux operating systems
+fn get_linux_response<S: Session>(session: &mut S, os_type: String) -> session::Result<()> {
+    #[cfg(target_os = "linux")]
+    {
+        let linux_release_info = linux_os_release()
+                .map_err(Error::CannotGetLinuxRelease)?;
+
+        let mut system_info: utsname;
+
+        unsafe {
+            system_info = std::mem::zeroed();
+            uname(&mut system_info);
+        }
+
+        session.reply(Response {
+            system: Some(os_type),
+            release_name: linux_release_info.name,
+            version_id: linux_release_info.version_id,
+            machine: Some(convert_raw_string(&system_info.machine)),
+            kernel_release: Some(convert_raw_string(&system_info.release)),
+            fqdn: hostname().ok(),
+            architecture: Some(convert_raw_string(&system_info.machine)),
+            node: Some(convert_raw_string(&system_info.nodename))
+        })?;
+    }
+    Ok(())
+}
+
+/// Handles requests for `GetPlatformInfo` action.
+pub fn handle<S: Session>(session:&mut S, _: ()) -> session::Result<()> {
+    let os_type = os_type().map_err(Error::CannotGetOSType)?;
+    match os_type.as_str() {
+        "Linux" => {
+            get_linux_response(session, os_type)?;
+        },
+        _ => {
+            session.reply(Response {
+                    system: Some(os_type),
+                    fqdn: hostname().ok(),
+                    ..Default::default() // TODO: Add other systems
+                })?;
+        }
+    }
+
+    Ok(())
+}
+
+impl super::Response for Response {
+
+    const RDF_NAME: Option<&'static str> = Some("Uname");
+
+    type Proto = rrg_proto::Uname;
+
+    /// Convert `Response` struct to protobuf message `Uname`
+    fn into_proto(self) -> rrg_proto::Uname {
+        Uname {
+            system: self.system.clone(),
+            release: self.release_name.clone(),
```

Comment:
Why do you need to clone? `into_proto` takes an owned response.

Reply:
I need to clone, because `pep425tag` also try to move these fields.

Reply:
Why? You are just stringifying them, borrowing them should be enough. You probably need to refactor it to a separate variable.